File: weka/pca.py
from subprocess import Popen, PIPE, STDOUT

# Settings to perform PCA.
classpath = '~/weka.jar'
filterClass = 'weka.filters.supervised.attribute.AttributeSelection'
searchClass = 'weka.attributeSelection.Ranker'
evaluationClass = 'weka.attributeSelection.PrincipalComponents -R 1.0 -C -A -1'
inputFile = '../data/ingredients.arff'

# Perform PCA.
p = Popen(['java', '-classpath', classpath,
			filterClass, 
			'-S', searchClass, '-E', evaluationClass, '-i', inputFile],
			shell=False, bufsize=1, stdout=PIPE, stderr=STDOUT, close_fds=True)

# Show transform data into eigenvector basis.
for line in iter(p.stdout.readline, b''):
    print line,
p.stdout.close()
p.wait()

File: weka/pca.java
import weka.core.Instances;
import weka.core.converters.ConverterUtils.DataSource;
import weka.attributeSelection.PrincipalComponents;
import weka.attributeSelection.Ranker;

public class pca {
	public static void main(String args[]) {
		try {
			// Load the Data.
			DataSource source = new DataSource("../data/ingredients.arff");
			Instances data = source.getDataSet();

			// Perform PCA.
			PrincipalComponents pca = new PrincipalComponents();
			pca.setVarianceCovered(1.0);
			pca.setCenterData(true);
			pca.setTransformBackToOriginal(false);
			pca.buildEvaluator(data);

			// Show transform data into eigenvector basis.
			Instances transformedData = pca.transformedData(data);
			System.out.println(transformedData);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

}
